File: backend/supabase/functions/_shared/calendarService.ts
import * as nse from "./nseService.ts";
import { daysBetween, indiaTodayStr } from "./wallclock.ts";

const REGULATORY_KEYWORDS = ["SEBI", "RAID", "RESIGNATION", "LOSS"];
const REGULATORY_PATTERNS = REGULATORY_KEYWORDS.map((k) => new RegExp(`\\b${k}\\b`, "i"));
const RESULTS_PATTERN = /\bresult/i;
const ANNOUNCEMENT_GRACE_DAYS = 5;

type CalendarEvent = {
  title: string;
  source: string;
  url: string;
  publishedAt: string | null;
  eventDate: string | null;
};

function quotePageUrl(symbol: string): string {
  return `https://www.nseindia.com/get-quotes/equity?symbol=${symbol}`;
}

const MONTHS: Record<string, number> = {
  Jan: 0, Feb: 1, Mar: 2, Apr: 3, May: 4, Jun: 5, Jul: 6, Aug: 7, Sep: 8, Oct: 9, Nov: 10, Dec: 11,
};

/** Parses NSE's "DD-Mon-YYYY" date format into "YYYY-MM-DD". */
function parseNseDate(value: string | null | undefined): string | null {
  if (!value || value === "-") return null;
  const m = /^(\d{1,2})-([A-Za-z]{3})-(\d{4})$/.exec(value.trim());
  if (!m) return null;
  const day = Number(m[1]);
  const month = MONTHS[m[2]];
  const year = Number(m[3]);
  if (month === undefined) return null;
  return `${year.toString().padStart(4, "0")}-${(month + 1).toString().padStart(2, "0")}-${day.toString().padStart(2, "0")}`;
}

/** Parses "DD-Mon-YYYY HH:MM:SS" (or falls back to date-only) into an ISO-ish naive string. */
function parseNseDateTime(value: string | null | undefined): string | null {
  if (!value) return null;
  const m = /^(\d{1,2})-([A-Za-z]{3})-(\d{4})\s+(\d{2}):(\d{2}):(\d{2})$/.exec(value.trim());
  if (m) {
    const day = Number(m[1]);
    const month = MONTHS[m[2]];
    const year = Number(m[3]);
    if (month === undefined) return null;
    return `${year.toString().padStart(4, "0")}-${(month + 1).toString().padStart(2, "0")}-${day.toString().padStart(2, "0")}T${m[4]}:${m[5]}:${m[6]}`;
  }
  const dateOnly = parseNseDate(value);
  return dateOnly ? `${dateOnly}T00:00:00` : null;
}

function event(title: string, url: string, eventDate: string | null, publishedAt: string | null = null): CalendarEvent {
  return { title, source: "NSE", url, publishedAt, eventDate };
}

async function buildMeetings(symbol: string, today: string): Promise<[CalendarEvent[], CalendarEvent[]]> {
  const meetings: CalendarEvent[] = [];
  const results: CalendarEvent[] = [];
  for (const bm of await nse.fetchBoardMeetings(symbol)) {
    const bmDate = parseNseDate(bm.bm_date);
    if (bmDate === null || bmDate < today) continue;
    const purpose = (bm.bm_purpose ?? "Board Meeting").trim();
    const desc = (bm.bm_desc ?? "").trim();
    const title = desc && desc.toLowerCase() !== purpose.toLowerCase() ? `${purpose} — ${desc}` : purpose;
    const entry = event(title, quotePageUrl(symbol), bmDate);
    meetings.push(entry);
    if (RESULTS_PATTERN.test(purpose) || RESULTS_PATTERN.test(desc)) results.push(entry);
  }
  return [meetings, results];
}

async function buildCorporateActions(symbol: string, today: string): Promise<CalendarEvent[]> {
  const actions: CalendarEvent[] = [];
  for (const ca of await nse.fetchCorporateActions(symbol)) {
    const eventDate = parseNseDate(ca.exDate) ?? parseNseDate(ca.recDate);
    if (eventDate === null || eventDate < today) continue;
    const subject = (ca.subject ?? "Corporate Action").trim();
    actions.push(event(subject, quotePageUrl(symbol), eventDate));
  }
  return actions;
}

async function buildRegulatory(symbol: string, today: string): Promise<CalendarEvent[]> {
  const flagged: CalendarEvent[] = [];
  for (const an of await nse.fetchAnnouncements(symbol)) {
    const haystack = `${an.desc ?? ""} ${an.attchmntText ?? ""}`;
    if (!REGULATORY_PATTERNS.some((p) => p.test(haystack))) continue;
    const filedAt = parseNseDateTime(an.an_dt);
    if (filedAt === null || daysBetween(today, filedAt.slice(0, 10)) > ANNOUNCEMENT_GRACE_DAYS) continue;
    const title = (an.attchmntText ?? an.desc ?? "Announcement").trim();
    const url = an.attchmntFile ?? quotePageUrl(symbol);
    flagged.push(event(title, url, null, filedAt));
  }
  return flagged;
}

export async function buildCalendarRow(scriptId: number, symbol: string): Promise<Record<string, unknown>> {
  const today = indiaTodayStr();
  const [meetings, results] = await buildMeetings(symbol, today);
  const corporateActions = await buildCorporateActions(symbol, today);
  const regulatory = await buildRegulatory(symbol, today);

  for (const bucket of [meetings, results, corporateActions]) {
    bucket.sort((a, b) => (a.eventDate ?? "").localeCompare(b.eventDate ?? ""));
  }
  regulatory.sort((a, b) => (b.publishedAt ?? "").localeCompare(a.publishedAt ?? ""));

  return {
    script_id: scriptId,
    symbol,
    meetings,
    corporate_actions: corporateActions,
    results,
    regulatory,
  };
}

export async function buildCalendar(scripts: [number, string][]): Promise<Record<string, unknown>[]> {
  return await Promise.all(scripts.map(([id, name]) => buildCalendarRow(id, name)));
}

File: backend/app/services/calendar_service.py
import re
from datetime import date, datetime

from app.services import nse_service

REGULATORY_KEYWORDS = ["SEBI", "RAID", "RESIGNATION", "LOSS"]
_REGULATORY_PATTERNS = [re.compile(rf"\b{re.escape(keyword)}\b", re.IGNORECASE) for keyword in REGULATORY_KEYWORDS]
_RESULTS_PATTERN = re.compile(r"\bresult", re.IGNORECASE)

# Announcement filings (unlike board meetings / corporate actions) carry no future
# event date of their own — keep a flagged one visible for a few days after filing,
# then let it drop off.
_ANNOUNCEMENT_GRACE_DAYS = 5


def _quote_page_url(symbol: str) -> str:
    return f"https://www.nseindia.com/get-quotes/equity?symbol={symbol}"


def _parse_nse_date(value: str | None) -> date | None:
    if not value or value == "-":
        return None
    try:
        return datetime.strptime(value.strip(), "%d-%b-%Y").date()
    except ValueError:
        return None


def _parse_nse_datetime(value: str | None) -> datetime | None:
    if not value:
        return None
    try:
        return datetime.strptime(value.strip(), "%d-%b-%Y %H:%M:%S")
    except ValueError:
        parsed_date = _parse_nse_date(value)
        return datetime.combine(parsed_date, datetime.min.time()) if parsed_date else None


def _event(title: str, url: str, event_date: date | None, published_at: datetime | None = None) -> dict:
    return {
        "title": title,
        "source": "NSE",
        "url": url,
        "publishedAt": published_at.isoformat() if published_at else None,
        "eventDate": event_date.isoformat() if event_date else None,
    }


def _build_meetings(symbol: str, today: date) -> tuple[list[dict], list[dict]]:
    meetings: list[dict] = []
    results: list[dict] = []
    for bm in nse_service.fetch_board_meetings(symbol):
        bm_date = _parse_nse_date(bm.get("bm_date"))
        if bm_date is None or bm_date < today:
            continue
        purpose = (bm.get("bm_purpose") or "Board Meeting").strip()
        desc = (bm.get("bm_desc") or "").strip()
        title = f"{purpose} — {desc}" if desc and desc.lower() != purpose.lower() else purpose
        entry = _event(title, _quote_page_url(symbol), bm_date)
        meetings.append(entry)
        if _RESULTS_PATTERN.search(purpose) or _RESULTS_PATTERN.search(desc):
            results.append(entry)
    return meetings, results


def _build_corporate_actions(symbol: str, today: date) -> list[dict]:
    actions: list[dict] = []
    for ca in nse_service.fetch_corporate_actions(symbol):
        event_date = _parse_nse_date(ca.get("exDate")) or _parse_nse_date(ca.get("recDate"))
        if event_date is None or event_date < today:
            continue
        subject = (ca.get("subject") or "Corporate Action").strip()
        actions.append(_event(subject, _quote_page_url(symbol), event_date))
    return actions


def _build_regulatory(symbol: str, today: date) -> list[dict]:
    flagged: list[dict] = []
    for an in nse_service.fetch_announcements(symbol):
        haystack = f"{an.get('desc') or ''} {an.get('attchmntText') or ''}"
        if not any(pattern.search(haystack) for pattern in _REGULATORY_PATTERNS):
            continue
        filed_at = _parse_nse_datetime(an.get("an_dt"))
        if filed_at is None or (today - filed_at.date()).days > _ANNOUNCEMENT_GRACE_DAYS:
            continue
        title = (an.get("attchmntText") or an.get("desc") or "Announcement").strip()
        url = an.get("attchmntFile") or _quote_page_url(symbol)
        flagged.append(_event(title, url, None, published_at=filed_at))
    return flagged


def build_calendar_row(script_id: int, symbol: str) -> dict:
    today = date.today()
    meetings, results = _build_meetings(symbol, today)
    corporate_actions = _build_corporate_actions(symbol, today)
    regulatory = _build_regulatory(symbol, today)

    for bucket in (meetings, results, corporate_actions):
        bucket.sort(key=lambda e: e["eventDate"] or "")
    regulatory.sort(key=lambda e: e["publishedAt"] or "", reverse=True)

    return {
        "script_id": script_id,
        "symbol": symbol,
        "meetings": meetings,
        "corporate_actions": corporate_actions,
        "results": results,
        "regulatory": regulatory,
    }


def build_calendar(scripts: list[tuple[int, str]]) -> list[dict]:
    return [build_calendar_row(script_id, symbol) for script_id, symbol in scripts]
